server/services/v1/cache/query_runner.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2022-2023 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/errors"
+	"github.com/tigrisdata/tigris/internal"
+	"github.com/tigrisdata/tigris/server/config"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/request"
+	"github.com/tigrisdata/tigris/server/services/v1/database"
+	"github.com/tigrisdata/tigris/server/transaction"
+	"github.com/tigrisdata/tigris/server/types"
+	"github.com/tigrisdata/tigris/store/cache"
+)
+
+type BaseQueryRunner struct {
+	encoder     metadata.CacheEncoder
+	cacheStore  cache.Cache
+	accessToken *types.AccessToken
+}
+
+func NewBaseQueryRunner(encoder metadata.CacheEncoder, accessToken *types.AccessToken, cacheStore cache.Cache) *BaseQueryRunner {
+	return &BaseQueryRunner{
+		encoder:     encoder,
+		accessToken: accessToken,
+		cacheStore:  cacheStore,
+	}
+}
+
+type CreateCacheQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.CreateCacheRequest
+}
+
+type DeleteCacheQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.DeleteCacheRequest
+}
+
+type SetQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.SetRequest
+}
+
+type GetQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.GetRequest
+}
+
+type DelQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.DelRequest
+}
+
+type KeysQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.KeysRequest
+}
+type QueryRunnerFactory struct {
+	encoder    metadata.CacheEncoder
+	cacheStore cache.Cache
+}
+
+// NewQueryRunnerFactory returns QueryRunnerFactory object.
+func NewQueryRunnerFactory(encoder metadata.CacheEncoder, cacheStore cache.Cache) *QueryRunnerFactory {
+	return &QueryRunnerFactory{
+		encoder:    encoder,
+		cacheStore: cacheStore,
+	}
+}
+
+func (f *QueryRunnerFactory) GetCreateCacheQueryRunner(r *api.CreateCacheRequest, accessToken *types.AccessToken) *CreateCacheQueryRunner {
+	return &CreateCacheQueryRunner{
+		BaseQueryRunner: NewBaseQueryRunner(f.encoder, accessToken, f.cacheStore),
+		req:             r,
+	}
+}
+
+func (f *QueryRunnerFactory) GetDeleteCacheQueryRunner(r *api.DeleteCacheRequest, accessToken *types.AccessToken) *DeleteCacheQueryRunner {
```

Comment:
Not related to this diff, but this pattern leads to many allocations per request.
The object is allocated on the heap, while it's only required for the duration of the request execution.

```
server/services/v1/api.go:463:50: &database.DatabaseQueryRunner{...} escapes to heap
server/services/v1/api.go:463:50: &database.BaseQueryRunner{...} escapes to heap
server/services/v1/api.go:466:47: &database.ReqOptions{} escapes to heap
```
@himank just to keep in mind, we may need to refactor it in the future.

Reply:
I think this pattern allows us a nice isolation/separation between executing different requests as we don't need to worry about shared state or contention and the runner if needed can cache anything for a lifetime of a request otherwise we need to think about all this if the object is shared. 

I am guessing you're more worried about allocation rather than this pattern itself? If yes, then there are a few things we can do,
- ReqOptions is to add our options, it is not needed on the heap. We can just pass it without worrying about passing the pointer to it. This will cut down the allocation by half. 
- DatabaseQueryRunner/BaseQueryRunner - this is a single object broken down into two to allow sharing some common code. Let's think about it may be we don't need it on the heap as well and just very well be on the stack itself.

Reply:
Right, mostly concerned about allocations.